tensorboard/plugins/audio/audio_plugin.py

```diff
@@ -178,78 +148,73 @@ def _audio_response_for_run(self, tensor_events, run, tag, sample):
-                    "query": self._query_for_individual_audio(
-                        run, tag, sample, index
-                    ),
+                    "query": query,
                 }
             )
         return response
 
-    def _query_for_individual_audio(self, run, tag, sample, index):
-        """Builds a URL for accessing the specified audio.
-
-        This should be kept in sync with _serve_audio_metadata. Note that the URL is
-        *not* guaranteed to always return the same audio, since audio may be
-        unloaded from the reservoir as new audio entries come in.
-
-        Args:
-          run: The name of the run.
-          tag: The tag.
-          index: The index of the audio entry. Negative values are OK.
-
-        Returns:
-          A string representation of a URL that will load the index-th sampled audio
-          in the given run with the given tag.
-        """
-        query_string = urllib.parse.urlencode(
-            {"run": run, "tag": tag, "sample": sample, "index": index,}
+    def _get_mime_type(self, experiment, run, tag):
+        # TODO(@wchargin): Move this call from `/audio` (called many
+        # times) to `/tags` (called few times) to reduce data provider
+        # calls.
+        self._data_provider.list_blob_sequences
+        mapping = self._data_provider.list_blob_sequences(
+            experiment_id=experiment, plugin_name=metadata.PLUGIN_NAME,
         )
-        return query_string
-
-    def _get_mime_type(self, run, tag):
-        content = self._multiplexer.SummaryMetadata(
-            run, tag
-        ).plugin_data.content
-        parsed = metadata.parse_plugin_metadata(content)
+        time_series = mapping.get(run, {}).get(tag, None)
+        if time_series is None:
+            raise errors.NotFoundError(
+                "No audio data for run=%r, tag=%r" % (run, tag)
+            )
+        parsed = metadata.parse_plugin_metadata(time_series.plugin_content)
         return _MIME_TYPES.get(parsed.encoding, _DEFAULT_MIME_TYPE)
 
     @wrappers.Request.application
     def _serve_individual_audio(self, request):
         """Serve encoded audio data."""
-        tag = request.args.get("tag")
-        run = request.args.get("run")
-        index = int(request.args.get("index", "0"))
-        sample = int(request.args.get("sample", "0"))
-        try:
-            events = self._filter_by_sample(
-                self._multiplexer.Tensors(run, tag), sample
-            )
-            data = events[index].tensor_proto.string_val[sample]
-        except (KeyError, IndexError):
-            return http_util.Respond(
-                request,
-                "Invalid run, tag, index, or sample",
-                "text/plain",
-                code=400,
+        experiment = plugin_util.experiment_id(request.environ)
+        mime_type = request.args["content_type"]
+        if mime_type not in _ALLOWED_MIME_TYPES:
+            raise errors.InvalidArgumentError(
```

Comment:
Thanks for catching this possible issue!  Can we add a test that exercises this to ensure that we refuse to return audio as, say, `text/javascript`?

I think the Werkzeug client `get()` API takes a `content_type` parameter so this should be straightforward.

In the longer term it would be nice if we didn't have to rely on the client behavior for this. Perhaps we should add a per-blob `content_type` onto the blob data model?  We would have to go around adding it to the various data providers and being sure to set it correctly on upload, but it seems worth doing to properly handle cases like this.

Reply:
Sure: done, I think, though I didn’t understand what you meant about the
Werkzeug client API taking a `content_type` parameter. (Isn’t that for
the content type of a POST request body?)

> Perhaps we should add a per-blob `content_type` onto the blob data
> model?

Perhaps. Presumably the source of truth would be a new field on the
`SummaryMetadata` proto (`string content_type = 5;`), required to be
homogeneous within a time series across all steps and indices, and
defaulting to `application/octet-stream` if empty for backward
compatibility? I’ll keep it in mind.

Reply:
Ah, I was confused and misread this and somehow thought we were getting the content type from a header that the browser was populating (even though yes, that isn't a thing for GET requests) rather than the query parameter.  Never mind :P 

Re: per-blob content types, yeah, I guess it could have a field in SummaryMetadata, although it's perhaps a little odd to have a field that is returned per-value stored in the per-tag SummaryMetadata (even if for now we don't anticipate allowing it to vary per-value).